### PR TITLE
implement JSON serialisation

### LINKED ISSUE #133: JSON serialzation
How to properly turn pydantic schemas into json?

The problem is that the BaseModel.dict() method returns dict, it's okay but it doesn't convert some default types into string (datetime.date, Enum and etc.).

It seems there is no other way to translate pydantic object into dict and serialize those types which is strange as pydantic can deserialize them and even force specific format in some way (datetime_parse.date_re for example which is hardcoded and hardly changeable).

### PR DESCRIPTION
**Pull request by samuelcolvin**
fix #133

### CODE REVIEW

File: tests/test_json.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    class Model(BaseModel):
+        a: float
+        b: bytes
+        c: Decimal
+        d: ModelA
+
+    m = Model(a=10.2, b='foobar', c=10.2, d={'x': 123, 'y': '123'})
+    assert m.dict() == {'a': 10.2, 'b': b'foobar', 'c': Decimal('10.2'), 'd': {'x': 123, 'y': '123'}}
+    assert m.json() == '{"a": 10.2, "b": "foobar", "c": 10.2, "d": {"x": 123, "y": "123"}}'
+    assert m.json(exclude={'b'}) == '{"a": 10.2, "c": 10.2, "d": {"x": 123, "y": "123"}}'
```
Comment (jasonkuhrt) [inline tests/test_json.py]: While its nice to have this call-site instance-level exclude flexibility, I think most use-cases will be based on standard `exclude`s at the definition level (aka. Class). Is it already possible to do that?

Comment (samuelcolvin) [reply tests/test_json.py]: no, for `dict` and `copy` exclude and include are provided here. I'd like to keep it the same.

Comment (jasonkuhrt) [reply tests/test_json.py]: Just thinking that _in addition_ to this there should be a way to specify this at the model definition level. In Go for instance they use annotations feature on structs to define how it maps to JSON when serialized.

Comment (jasonkuhrt) [reply tests/test_json.py]: And actually this is effectively a new feature so should be a new issue/pr etc. anyways... carry on 😅

Comment (samuelcolvin) [reply tests/test_json.py]: interesting, but as you say, definitely separate from this.

File: pydantic/json.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    frozenset: list,
+    GeneratorType: list,
+    bytes: lambda o: o.decode(),
+    Decimal: float,
+}
+
+
+def pydantic_encoder(obj):
+    if isinstance(obj, BaseModel):
+        return obj.dict()
```
Comment (jasonkuhrt) [inline pydantic/json.py]: A Python newb question: Why not `dict(obj)` here?

Comment (samuelcolvin) [reply pydantic/json.py]: no difference, models are setup so `dict(obj) == obj.dict()`. But generally I use this way.

**Pull request merged**